filter_sarif.py

```diff
@@ -1,147 +1,218 @@
+    """Match a path, rule and message against a list of patterns."""
     result = True
-    for s, fp, rp in patterns:
-        if match(rp, rule) and match(fp, path):
-            result = s
+
+    LOG.debug("Matching patterns against %s:%s:%s", path, rule, message)
+
+    for sign, file_pattern, rule_pattern, message_pattern in patterns:
+        LOG.debug("Matching against %s", (sign, file_pattern, rule_pattern, message_pattern))
+        if (path is None or file_pattern is None or
+                match(file_pattern, path)) and (rule is None or rule_pattern is None or match(rule_pattern, rule)) and (
+                    message_pattern is None or message is None or re.search(message_pattern, message)):
+            LOG.debug("Matched")
+            result = sign
+
     return result
 
 
-def parse_pattern(line):
-    sepchar = ':'
-    escchar = '\\'
-    file_pattern = ''
-    rule_pattern = ''
-    seen_separator = False
+def parse_pattern(line: str) -> Tuple[bool, str, Optional[str], Optional[str]]:
+    """Parse a pattern line into a tuple of (sign, file_pattern, rule_pattern, message_pattern)."""
     sign = True
+    file_pattern: str = ''
+    rule_pattern: Optional[str] = None
+    message_pattern: Optional[str] = None
+    seen_separator: int = 0
 
     # inclusion or exclusion pattern?
     uline = line
     if line:
-        if line[0] == '-':
+        if line[0] == exclude_char:
             sign = False
             uline = line[1:]
-        elif line[0] == '+':
+        elif line[0] == include_char:
             uline = line[1:]
 
+    # parse patterns character by character
     i = 0
-    while i < len(uline):
+    line_length = len(uline)
+
+    while i < line_length:
         c = uline[i]
-        i = i + 1
+        i += 1
         if c == sepchar:
-            if seen_separator:
-                raise Exception('Invalid pattern: "' + line + '" Contains more than one separator!')
-            seen_separator = True
+            seen_separator += 1
+            if seen_separator > 2:
+                raise Exception(
+                    f'Invalid pattern: "{line}" contains more than two separators. Use {escchar}{sepchar} to escape the separator.'
+                )
+            if seen_separator == 1:
+                rule_pattern = ''
+            elif seen_separator == 2:
+                message_pattern = ''
             continue
         elif c == escchar:
-            nextc = uline[i] if (i < len(uline)) else None
-            if nextc in ['+' , '-', escchar, sepchar]:
-                i = i + 1
+            nextc = uline[i] if (i < line_length) else None
+            if nextc in special_chars:
+                i += 1
                 c = nextc
-        if seen_separator:
-            rule_pattern = rule_pattern + c
+        if seen_separator == 1:
+            rule_pattern = rule_pattern + c if rule_pattern is not None else None
```

Comment:
Is this check actually required here? I would think at this point `rule_pattern` is never None, or am I overlooking something?
Also, can we have parens around that addition? The operator precedence is a bit unexpected here, at least for me as a casual Python user.

Reply:
This whole method is becoming quite complex, perhaps we could just use regular expressions instead:
```python
def parse_pattern(line: str) -> Tuple[bool, str, Optional[str], Optional[str]]:
  p = r'((?:[^:\\]|\\.)+)'       # if we want to allow empty patterns, we might want to use '*' instead
  m = re.fullmatch(f'([+-]?){p}(?::{p}(?::{p})?)?', line, re.DOTALL)
  if m:
    return (
      True if m.group(1) != '-' else False,
      *map(lambda g: re.sub(r'\\([+-\\:])', r'\1', g) if g else g, m.groups(None)[1:])
    )
  raise Exception("Syntax error!")
```

Reply:
Sure, I considered that (and would turn to that first, over a state machine) but wanted to stick with your parsing. Doing regex parsing to try to deal with escaping can tend to be a bit hard to write/read, but I like your solution.

Reply:
There is a `None` check needed, since the `rule` and `message` can be be empty and not parsed/populated by the state machine.

It's also needed to pass `MyPy` type checking, and sends a signal to anyone altering the state machine about the known state of the variables at that time.

Reply:
> There is a None check needed, since the rule and message can be be empty and not parsed/populated by the state machine.
Ah ok, it looked like that was being taken care of by lines 78-81, but I might be wrong. I must admit I didn't run the code yet.

Reply:
I've left this logic as-is. You're right that the state machine logic will guarantee that the rule/message are not None at that point, but `mypy` isn't clever enough to know that. The `None` check is just for the sake of that.